Assignment5/Q1.py

```diff
@@ -0,0 +1,150 @@
+them) and M is the number of dictionary words. There are two main steps performed in the given solution. First is to
+create a graph corresponding to the partial (lexicographical) order of characters in dictionary words. Second step
+creates a total order of characters (i. e. the lexicographical order or in other words ordered alphabet) which are
+stored at first in a singly linked list and then copied to the list of characters, This step uses a topological sort
+algorithm."""
+
+import SinglyLinkedList
+
+
+def get_index_of_first_differing_char(word1, word2):
+    """The method returns an index of the first character which is different in two given words.
+
+        Args:
+            word1: a string, the first word to perform comparison on
+            word2: a string, the second word to perform comparison on
+
+        Returns:
+            an integer, the index of first character which is different in two words, if the words
+            are the same or the beginning of longer word is the same as the whole shorter word, it
+            returns -1.
+    """
+    # Set length to the length of shorter word.
+    length = len(word1) if len(word1) < len(word2) else len(word2)
+    index = -1
+
+    # Loop over words, after loop is executed index will point to last character that is the same in two words.
+    for i in range(length):
+        if word1[i] == word2[i]:
+            index += 1
+        else:
+            break
+
+    # Point index to the first differing character in two words.
+    index += 1
+
+    # Reset index to -1 if the words are the same or beginning of longer word is the same as the whole shorter word.
+    if index == length:
+        index = -1
+
+    return index
+
+
+def create_corresponding_partial_order(dictionary):
+    """The method returns a partial order of characters from dictionary, based on the order of words in a dictionary,
+       in the graph form.
+
+        Args:
+            dictionary: an array of strings, list of language words in lexicographical order, the character partial
+                        ordering is created based on it
+
+        Returns:
+            graph: a dictionary of sets, corresponds to graph data structure, where characters from dictionary are
+                   'vertices' and 'edges' indicate characters' partial ordering
+            state: a dictionary of strings, where each string is either "white", "grey" or "black". It indicates the
+                   state that the "vertex" (character) is in while topological search is performed. In this method
+                   only initialize state dictionary to all "white".
+    """
+    graph = {}
+    state = {}
+
+    # Instantiate the graph data structure and set state to "white" for all characters/,
+    for word in dictionary:
+        for char in word:
+            if char not in graph:
+                graph[char] = set()
+                state[char] = "white"
```

Comment:
Instead of using strings, numbers would be more efficient. If you want, you can use constants like:
```python
_WHITE = 0
_GREY = 1
_BLACK = 2
```
to make it more explicit.